pkg/simple/client/monitoring/metricsserver/metricsserver.go

```diff
@@ -0,0 +1,386 @@
+
+	edgeNodeNamesFiltered := intersect(edgeNodeNames, nodeNames)
+
+	return edgeNodeNamesFiltered
+}
+
+func (m metricsserver) getNodeMetricsFromMetricsAPI() (*metricsapi.NodeMetricsList, error) {
+	var err error
+	versionedMetrics := &metricsV1beta1api.NodeMetricsList{}
+	mc := m.metricsClientSet.MetricsV1beta1()
+	nm := mc.NodeMetricses()
+	versionedMetrics, err = nm.List(context.TODO(), metav1.ListOptions{LabelSelector: edgeNodeLabel})
+	if err != nil {
+		return nil, err
+	}
+	metrics := &metricsapi.NodeMetricsList{}
+	err = metricsV1beta1api.Convert_v1beta1_NodeMetricsList_To_metrics_NodeMetricsList(versionedMetrics, metrics, nil)
+	if err != nil {
+		return nil, err
+	}
+	return metrics, nil
+}
+
+func NewMetricsServer(k kubernetes.Interface, options *k8s.KubernetesOptions) (monitoring.Interface, error) {
+	var metricsServer metricsserver
+	config, err := clientcmd.BuildConfigFromFlags("", options.KubeConfig)
+	if err != nil {
+		klog.Error(err)
+		return metricsServer, err
+	}
+
+	metricsServer.k8s = k
+
+	discoveryClient := k.Discovery()
+	apiGroups, err := discoveryClient.ServerGroups()
+	if err != nil {
+		klog.Error(err)
+		return metricsServer, err
+	}
+
+	metricsServer.metricsAPIAvailable = supportedMetricsAPIVersionAvailable(apiGroups)
+
+	if !metricsServer.metricsAPIAvailable {
+		klog.Warningf("Metrics API not available.")
+		return metricsServer, err
+	}
+
+	metricsClient, err := metricsclientset.NewForConfig(config)
+	if err != nil {
+		klog.Error(err)
+		return metricsServer, err
+	}
+
+	metricsServer.metricsClientSet = metricsClient
+	return metricsServer, nil
+}
+
+func (m metricsserver) GetMetric(expr string, ts time.Time) monitoring.Metric {
+	var parsedResp monitoring.Metric
+
+	return parsedResp
+}
+
+func (m metricsserver) GetMetricOverTime(expr string, start, end time.Time, step time.Duration) monitoring.Metric {
+	var parsedResp monitoring.Metric
+
+	return parsedResp
+}
+
+var edgeNodeMetrics = []string{"node_cpu_usage", "node_cpu_total", "node_cpu_utilisation", "node_memory_usage_wo_cache", "node_memory_total", "node_memory_utilisation"}
+
+func (m metricsserver) GetNamedMetrics(metrics []string, ts time.Time, o monitoring.QueryOption) []monitoring.Metric {
+	var res []monitoring.Metric
+
+	opts := monitoring.NewQueryOptions()
+	o.Apply(opts)
+	if opts.Level == monitoring.LevelNode {
+		edgeNodes, err := m.listEdgeNodes()
+		if err != nil {
+			klog.Errorf("List edge nodes error %v\n", err)
+			return res
+		}
+
+		if !m.metricsAPIAvailable {
+			klog.Warningf("Metrics API not available.")
+			return res
+		}
+
+		edgeNodeNamesFiltered := m.filterEdgeNodeNames(edgeNodes, opts)
+		if len(edgeNodeNamesFiltered) == 0 {
+			klog.V(4).Infof("No edge node metrics is requested")
+			return res
+		}
+		edgeNodeNamesMap := make(map[string]struct{}, len(edgeNodeNamesFiltered))
+		for _, s := range edgeNodeNamesFiltered {
+			edgeNodeNamesMap[s] = struct{}{}
+		}
+
+		metricsResult, err := m.getNodeMetricsFromMetricsAPI()
```

Comment:
Can we return a map[nodename]NodeMetrics{} in getNodeMetricsFromMetricsAPI ? 
It will more efficient to iterating over filteredNodeList and use node name to find the NodeMetrics .

Reply:
getNodeMetricsFromMetricsAPI still returns a list, but next loop uses edgeNodeNamesFiltered as map, so complexity is reduced to O(n), and Convert_v1beta1_NodeMetricsList_To_metrics_NodeMetricsList returns a list, converting it to map still need one full loop, so use metricsResult as list and edgeNodeNamesFiltered  as map is a proper choice.